### PR TITLE
[v8r0] Implement a recursive listDirectory for the DFC

### PR DESCRIPTION
**Pull request by chaen**
closes https://github.com/DIRACGrid/DIRAC/issues/3911

I'd have hoped for a simple flag in the `listDirectory` method, but it unfortunately can't be done (too much modification of too many methods that are used by too many components). 

So I created a new `getDirectoryDump` method. For a given directory, it returns all the files and sub folders contained at any depth. 

For now I've applied it only to the `dirac-dms-user-lfn` commands. In the case of the LHCb managers, the speedup is substantial. 

For example, on the test instance, I can dump 250k files and 100k sub directories in less than 10 seconds. The original command never returned....

```bash
time dirac-dms-user-lfns -b /lhcb/user/c/cburr/
Will search for files in /lhcb/user/c/cburr
/lhcb/user/c/cburr: 251681 files, 98570 sub-directories
251681 matched files have been put in lhcb-user-c-cburr.lfns

real    0m9.893s
user    0m4.280s
sys     0m0.248s
``` 

BEGINRELEASENOTES

*DMS
NEW: Introduce getDirectoryDump method in the DFC
CHANGE: dirac-dms-user-lfns uses getDirectoryDump method
CHANGE: DataManager getReplicasFromDirectory and __cleanDirectory use getDirectoryDump 

ENDRELEASENOTES

### CODE REVIEW

File: tests/Integration/DataManagementSystem/Test_FileCatalogDB.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 from DIRAC.Core.Security.Properties import FC_MANAGEMENT
 
+# CHRIS
+
```
Comment (chrisburr) [inline tests/Integration/DataManagementSystem/Test_FileCatalogDB.py]:
```suggestion
```

File: src/DIRAC/DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql [hunk 2]
```diff
@@ -493,6 +493,9 @@ DELIMITER ;
 -- includeParent: if true, include oneself
 -- returns (directory id, absolute level)
 
+-- CHRIS: CAN THIS BE REPLACED WITH SOMETHING LIKE
```
Comment (chrisburr) [inline src/DIRAC/DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql:496]: I guess this is a long-term reminder?

File: src/DIRAC/DataManagementSystem/Client/DataManager.py [hunk 3]
```diff
@@ -160,24 +160,22 @@ def __cleanDirectory(self, folder):
...
+
+        res = returnSingleResult(self.fileCatalog.getDirectoryDump([folder]))
+
         if not res["OK"]:
             return res
 
-        if not res["Value"]:
+        if not (res["Value"]["Files"] or res["Value"]["SubDirs"]):
             # folder is empty, just remove it and return
             res = returnSingleResult(self.fileCatalog.removeDirectory(folder, recursive=True))
```
Comment (chrisburr) [inline src/DIRAC/DataManagementSystem/Client/DataManager.py:171]: I'm surprised that this has `recursive=True`?

File: src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryClosure.py [hunk 4]
```diff
@@ -657,3 +658,40 @@ def _changeDirectoryParameter(self, paths, directoryFunction, _fileFunction, rec
                 successful[path] = True
 
         return S_OK({"Successful": successful, "Failed": failed})
+
+    def _getDirectoryDump(self, path):
+        """For a given directory, dump all the content of a directory
+
+        :param int dirID: directory ID
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryClosure.py]: none of these `:param` exist

File: src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py [hunk 5]
```diff
@@ -676,6 +677,89 @@ def listDirectory(self, lfns, verbose=False):
...
+
+        return S_OK({"Successful": successful, "Failed": failed})
+
+    def _getDirectoryDump(self, path):
+        """
+        Recursively dump all the content of a directory
+
+        :param str path: directory to dump
+
+        :returns: dictionary with `files` and `SubDirs` as keys
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py]:
```suggestion
        :returns: dictionary with `Files` and `SubDirs` as keys
```

File: src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py [hunk 6]
```diff
@@ -676,6 +677,89 @@ def listDirectory(self, lfns, verbose=False):
...
+                  `files` is a dict containing files metadata.
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py]:
```suggestion
                  `Files` is a dict containing files metadata.
```

File: src/DIRAC/DataManagementSystem/Service/FileCatalogHandler.py [hunk 7]
```diff
@@ -351,6 +351,12 @@ def export_getDirectoryReplicas(self, lfns, allStatus=False):
         """Get replicas for files in the supplied directory"""
         return self.fileCatalogDB.getDirectoryReplicas(lfns, allStatus, self.getRemoteCredentials())
 
+    types_getDirectoryDump = [[list, dict, str]]
+
+    def export_getDirectoryDump(self, lfns):
+        """List the contents of supplied directories"""
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Service/FileCatalogHandler.py]:
```suggestion
        """Recursively list the contents of supplied directories"""
```

File: src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py [hunk 8]
```diff
@@ -676,6 +677,89 @@ def listDirectory(self, lfns, verbose=False):
...
+                    {
+                        os.path.join(dirName, fileName): {
+                            "Size": fileMetadata["MetaData"]["Size"],
+                            "CreationDate": fileMetadata["MetaData"]["CreationDate"],
+                        }
+                        for fileName, fileMetadata in filesInDir.items()
+                    }
+                )
+
+            # Get subdirectories
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryTreeBase.py]:
```suggestion
            # Add to this list to get subdirectories of these directories
```

File: src/DIRAC/DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql (hunk 2 above)
Comment (chaen) [reply src/DIRAC/DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql:496]: Yes :D

File: src/DIRAC/DataManagementSystem/Client/DataManager.py (hunk 3 above)
Comment (chaen) [reply src/DIRAC/DataManagementSystem/Client/DataManager.py:171]: It's maybe not needed indeed, but I'd rather not change it here in the risk of breaking something, and I'd like this to be merged before Monday's release :angel:

File: src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryClosure.py (hunk 4 above)
Comment (chaen) [reply src/DIRAC/DataManagementSystem/DB/FileCatalogComponents/DirectoryManager/DirectoryClosure.py]: Oops, thanks :-)

**Review by chrisburr — APPROVED** [review]

**Pull request merged**